ibis/expr/operations.py

```diff
@@ -12,25 +12,154 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import operator
-import six
-import itertools
 
+import six
 import toolz
+import operator
+import itertools
+import collections
 
-from ibis.expr.types import TableColumn  # noqa
-
+from functools import partial
 from ibis.expr.schema import HasSchema, Schema
-from ibis.expr.rules import value, string, number, integer, boolean, list_of
-from ibis.expr.types import (Node, as_value_expr, Expr,
-                             ValueExpr, ColumnExpr, TableExpr,
-                             ValueOp, _safe_repr, distinct_roots)
-import ibis.common as com
-import ibis.expr.datatypes as dt
-import ibis.expr.rules as rules
-import ibis.expr.types as ir
+
 import ibis.util as util
+import ibis.common as com
 import ibis.compat as compat
+import ibis.expr.types as ir
+import ibis.expr.rules as rlz
+import ibis.expr.schema as sch
+import ibis.expr.datatypes as dt
+from ibis.expr.signature import Annotable, Argument as Arg
+
+
+def _safe_repr(x, memo=None):
+    return x._repr(memo=memo) if isinstance(x, (ir.Expr, Node)) else repr(x)
+
+
+# TODO: move to analysis
+def distinct_roots(*expressions):
+    roots = toolz.concat(
+        expression._root_tables() for expression in expressions
+    )
+    return list(toolz.unique(roots, key=id))
+
+
+class Node(Annotable):
+
+    __slots__ = ('_expr_cached',)
+
+    def __repr__(self):
+        return self._repr()
+
+    def _repr(self, memo=None):
+        if memo is None:
+            from ibis.expr.format import FormatMemo
```

Comment:
Probably ok to just import in the function, this is a little strange.

Reply:
There are circular imports.